File: surface/widgets/compose/components/widgets/domain/Minimap.kt
// ============================================================
// Clef Surface Compose Widget — Minimap
//
// Scaled-down overview of a larger document or canvas, rendered
// as a small Canvas composable. Uses density shading to represent
// content and highlights the current viewport position within
// the full document.
//
// Adapts the minimap.widget spec: anatomy (root, canvas,
// viewport, zoomControls, zoomIn, zoomOut, zoomFit, zoomLevel),
// states (idle, panning), and connect attributes.
// ============================================================

package clef.surface.compose.components.widgets.domain

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// --------------- Helpers ---------------

private fun computeDensity(line: String): Float {
    val trimmed = line.trim()
    if (trimmed.isEmpty()) return 0f
    return (trimmed.length.toFloat() / line.length.coerceAtLeast(1)).coerceIn(0f, 1f)
}

// --------------- Component ---------------

/**
 * Small Canvas overview minimap of document content.
 *
 * @param content Lines of content to visualize.
 * @param visibleStart First visible line index.
 * @param visibleEnd Last visible line index (exclusive).
 * @param totalLines Total number of lines in the document.
 * @param width Width of the minimap.
 * @param height Height of the minimap.
 * @param modifier Modifier for the root layout.
 */
@Composable
fun Minimap(
    content: List<String>,
    visibleStart: Int,
    visibleEnd: Int,
    totalLines: Int,
    width: Dp = 80.dp,
    height: Dp = 120.dp,
    modifier: Modifier = Modifier,
) {
    val densities = remember(content, totalLines, height) {
        if (totalLines == 0) return@remember emptyList<Float>()
        val rowCount = 60 // virtual rows to sample
        val linesPerRow = (totalLines.toFloat() / rowCount).coerceAtLeast(1f)
        (0 until rowCount).map { row ->
            val sampleLine = (row * linesPerRow).toInt().coerceAtMost(content.lastIndex.coerceAtLeast(0))
            computeDensity(content.getOrElse(sampleLine) { "" })
        }
    }

    Column(
        modifier = modifier.padding(4.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Canvas(
            modifier = Modifier
                .size(width, height)
                .border(1.dp, MaterialTheme.colorScheme.outline),
        ) {
            if (densities.isEmpty() || totalLines == 0) return@Canvas

            val rowHeight = size.height / densities.size

            // Draw density bars
            densities.forEachIndexed { index, density ->
                if (density > 0f) {
                    val barWidth = size.width * density
                    drawRect(
                        color = Color.Gray.copy(alpha = 0.2f + density * 0.4f),
                        topLeft = Offset(0f, index * rowHeight),
                        size = Size(barWidth, rowHeight),
                    )
                }
            }

            // Draw viewport indicator
            val viewTopRatio = visibleStart.toFloat() / totalLines
            val viewBottomRatio = visibleEnd.toFloat() / totalLines
            val viewTop = viewTopRatio * size.height
            val viewHeight = (viewBottomRatio - viewTopRatio) * size.height

            drawRect(
                color = Color(0xFF6200EE).copy(alpha = 0.3f),
                topLeft = Offset(0f, viewTop),
                size = Size(size.width, viewHeight),
            )
            drawRect(
                color = Color(0xFF6200EE),
                topLeft = Offset(0f, viewTop),
                size = Size(size.width, viewHeight),
                style = Stroke(width = 2f),
            )
        }

        Spacer(modifier = Modifier.height(4.dp))
        Text(
            text = "${visibleStart + 1}-$visibleEnd/$totalLines",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
    }
}

File: surface/widgets/swiftui/components/widgets/domain/Minimap.swift
// ============================================================
// Clef Surface SwiftUI Widget — Minimap
//
// Scaled-down overview of a larger document or canvas, rendered
// as a small Canvas. Uses density shading to represent content
// and highlights the current viewport position within the full
// document.
// ============================================================

import SwiftUI

private func computeDensity(_ line: String) -> CGFloat {
    let trimmed = line.trimmingCharacters(in: .whitespaces)
    guard !trimmed.isEmpty else { return 0 }
    return min(CGFloat(trimmed.count) / max(CGFloat(line.count), 1), 1)
}

struct MinimapView: View {
    var content: [String]
    var visibleStart: Int
    var visibleEnd: Int
    var totalLines: Int
    var minimapWidth: CGFloat = 80
    var minimapHeight: CGFloat = 120

    var body: some View {
        let densities: [CGFloat] = {
            guard totalLines > 0 else { return [] }
            let rowCount = 60
            let linesPerRow = max(CGFloat(totalLines) / CGFloat(rowCount), 1)
            return (0..<rowCount).map { row in
                let sampleLine = min(Int(CGFloat(row) * linesPerRow), max(content.count - 1, 0))
                return computeDensity(sampleLine < content.count ? content[sampleLine] : "")
            }
        }()

        VStack(spacing: 4) {
            Canvas { context, size in
                guard !densities.isEmpty, totalLines > 0 else { return }

                let rowHeight = size.height / CGFloat(densities.count)

                // Draw density bars
                for (index, density) in densities.enumerated() {
                    if density > 0 {
                        let barWidth = size.width * density
                        let rect = CGRect(x: 0, y: CGFloat(index) * rowHeight, width: barWidth, height: rowHeight)
                        context.fill(
                            Rectangle().path(in: rect),
                            with: .color(Color.gray.opacity(0.2 + Double(density) * 0.4))
                        )
                    }
                }

                // Draw viewport indicator
                let viewTopRatio = CGFloat(visibleStart) / CGFloat(totalLines)
                let viewBottomRatio = CGFloat(visibleEnd) / CGFloat(totalLines)
                let viewTop = viewTopRatio * size.height
                let viewHeight = (viewBottomRatio - viewTopRatio) * size.height
                let viewRect = CGRect(x: 0, y: viewTop, width: size.width, height: viewHeight)

                context.fill(
                    Rectangle().path(in: viewRect),
                    with: .color(Color.purple.opacity(0.3))
                )
                context.stroke(
                    Rectangle().path(in: viewRect),
                    with: .color(.purple),
                    style: StrokeStyle(lineWidth: 2)
                )
            }
            .frame(width: minimapWidth, height: minimapHeight)
            .border(Color(.systemGray3), width: 1)

            Text("\(visibleStart + 1)-\(visibleEnd)/\(totalLines)")
                .font(.caption2)
                .foregroundColor(.secondary)
        }
        .padding(4)
    }
}
